### PR TITLE
Optimize get block by evm number

### PR DESCRIPTION
**Pull request by lochjin**

### CODE REVIEW

File: core/blockchain/blockindex.go [hunk 1]
```diff
@@ -191,3 +191,57 @@ func (b *BlockChain) GetMiningTips(expectPriority int) []*hash.Hash {
 func (b *BlockChain) HasTx(txid *hash.Hash) bool {
 	return b.indexManager.HasTx(txid)
 }
+
+func (b *BlockChain) GetBlockByNumber(number uint64) meerdag.IBlock {
```
Comment (dindinw) [inline core/blockchain/blockindex.go:195]: code comments are required

**Review by dindinw — CHANGES_REQUESTED** [review]
In general, It's okey for me to remove the `vmblockindex` entirely since the recent code refactoing.  but the search performance of the `evm_tx_id` -> `dag_tx_id` is very different than before cause the usage of the new added function `get_block_by_number`, which is implemented by searching the entire block dag using the binary search. 
I think the `evm <-> blockdag` index might be still required in the future.

File: services/tx/api.go [hunk 2]
```diff
@@ -1003,15 +1002,11 @@ func (api *PublicTxAPI) GetTxIDByMeerEVMTxHash(etxh hash.Hash) (interface{}, err
 	if bid == 0 {
 		return nil, fmt.Errorf("No meerevm tx:%s", etxh.String())
 	}
-	vmbiStore := api.txManager.consensus.VMBlockIndexStore()
-	if vmbiStore == nil {
-		return nil, fmt.Errorf("You must be enable by --vmblockindex")
+	b := api.txManager.GetChain().GetBlockByNumber(bid)
```
Comment (dindinw) [inline services/tx/api.go:1005]: bid is a evm block number?

Comment (lochjin) [reply services/tx/api.go:1005]: Yes, this `bid` came from EVM API

File: core/blockchain/api.go [hunk 3]
```diff
@@ -505,3 +372,87 @@ func (api *PublicBlockAPI) GetStateRoot(order int64, verbose *bool) (interface{}
 	}
 	return ib.GetState().Root().String(), nil
 }
+
+func (api *PublicBlockAPI) GetBlockByNumber(number uint64, verbose *bool, inclTx *bool, fullTx *bool) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (lochjin) [reply core/blockchain/api.go:380]: https://github.com/Qitmeer/qng/pull/425/commits/2a9ca35614056152e9eaec9b398721db71f5a88f

File: core/blockchain/blockindex.go (hunk 1 above)
Comment (lochjin) [reply core/blockchain/blockindex.go:197]: https://github.com/Qitmeer/qng/pull/425/commits/00aa08a77f110700cfc4f8ce28513fe3c514d27e

**Comment by lochjin** [comment]
> In general, It's okey for me to remove the `vmblockindex` entirely since the recent code refactoing. but the search performance of the `evm_tx_id` -> `dag_tx_id` is very different than before cause the usage of the new added function `get_block_by_number`, which is implemented by searching the entire block dag using the binary search.
> I think the `evm <-> blockdag` index might be still required in the future.

Yes, the current implementation has performance bottlenecks. Future needs to be replaced with more efficient solutions

**Pull request merged**